File: 00951. Flip Equivalent Binary Trees/solution.go
/**
 * Problem: 951. Flip Equivalent Binary Trees
 *
 * Difficulty: Medium
 *
 * Language: Go
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

func flipEquiv(root1 *TreeNode, root2 *TreeNode) bool {
	// Check if both trees are empty - they're equivalent
	if root1 == nil && root2 == nil {
		return true
	}
	// Check if only one tree is empty - they can't be equivalent
	if root1 == nil || root2 == nil {
		return false
	}
	// Check if node values are different - they can't be equivalent
	if root1.Val != root2.Val {
		return false
	}
	
	// Try without flipping - check if subtrees match in normal order
	noFlip := flipEquiv(root1.Left, root2.Left) && flipEquiv(root1.Right, root2.Right)
	// Try with flipping - check if subtrees match when swapped
	flip := flipEquiv(root1.Left, root2.Right) && flipEquiv(root1.Right, root2.Left)
	
	// Trees are equivalent if either normal order or flipped order works
	return noFlip || flip
}

File: 00951. Flip Equivalent Binary Trees/solution.js
/**
 * Problem: 951. Flip Equivalent Binary Trees
 *
 * Difficulty: Medium
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

/**
 * Determines if two binary trees are flip equivalent (can become identical by flipping subtrees)
 *
 * @param {TreeNode} root1 - First tree root
 * @param {TreeNode} root2 - Second tree root
 *
 * @returns {boolean} - True if trees are flip equivalent
 */
const flipEquiv = (root1, root2) => {
  // Return true if both nodes are null, false if only one is null
  if (!root1 || !root2) return !root1 && !root2
  // Return false if values don't match (can't be flip equivalent)
  if (root1.val !== root2.val) return false

  // Try without flipping - check if left-left and right-right match
  const noFlip =
    flipEquiv(root1.left, root2.left) && flipEquiv(root1.right, root2.right)
  // Try with flipping - check if left-right and right-left match
  const withFlip =
    flipEquiv(root1.left, root2.right) && flipEquiv(root1.right, root2.left)

  // Trees are flip equivalent if either configuration works
  return noFlip || withFlip
}
